### ISSUE TITLE
Installation issue: openbabel

### ISSUE THREAD

**Issue by omsai** [LLNL/spack#18878]
Open Babel 3.0.0 and the latest 3.1.1 automagically detects the system wxwidgets and link to GTK2 libraries on RedHat 6, which causes the fun failures below.  This can be disabled in upstream's CMake via BUILD_GUI=OFF.  Good practice in packaging is not letting these options up to chance and having the wrapper - spack's variants - control them.  But I can't think why someone would want the openbabel GUI on HPC, so it might make sense to simply hard code disable the GUI.

### Steps to reproduce the issue

```console
$ spack install openbabel
...
==> Installing openbabel                                                                                                 
==> No binary for openbabel found: installing from source                                                                  
Fetching tags only, you probably meant:                                                                                   
  git fetch --tags                                                                                                            
==> openbabel: Executing phase: 'cmake'                                        
==> openbabel: Executing phase: 'build'                                                        
==> Error: ProcessError: Command exited with status 2:                                                                                                            
    'make' '-j12'                                                                              
                                                                                               
10 errors found in build log:                                                  
...
  >> 2658    /usr/lib64/libgdk-x11-2.0.so.0: undefined reference to `cairo_xlib_surface_create_for_bitmap'                                                       
  >> 2659    /usr/lib64/libgdk-x11-2.0.so.0: undefined reference to `cairo_xlib_surface_create'                                                  
  >> 2660    /usr/lib64/libgtk-x11-2.0.so.0: undefined reference to `cairo_pdf_surface_set_size'                                                 
  >> 2661    /usr/lib64/libgtk-x11-2.0.so.0: undefined reference to `cairo_pdf_surface_create_for_stream'                                                                                    
  >> 2662    /usr/lib64/libgtk-x11-2.0.so.0: undefined reference to `cairo_pdf_surface_create'                                                   
  >> 2663    /usr/lib64/libgdk-x11-2.0.so.0: undefined reference to `cairo_xlib_surface_set_size'                                                        
  >> 2664    collect2: error: ld returned 1 exit status                        
  >> 2665    make[2]: *** [bin/obgui] Error 1                                  
     2666    make[2]: Leaving directory `/tmp/pan14001admin/spack-stage/spack-stage-openbabel-3.0.0-mcjgl4w4ermkcy5xj3ras2ovwvooj7gh/spack-build=mcjgl4w'
```

### Information on your system

```console
$ spack debug report
* **Spack:** 0.15.4-1045-678e9f5
* **Python:** 3.4.5
* **Platform:** linux-rhel6-westmere
```

```yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs:
  - librsvg@2.50.0 ^cairo +pdf
  - openbabel@3.1.1
  packages:
    autoconf:
      buildable: false
      externals:
      - spec: autoconf@2.69
        prefix: /apps2/autoconf/2.69
      target: []
      compiler: []
      version: []
      providers: {}
    automake:
      buildable: false
      externals:
      - spec: automake@1.16
        prefix: /apps2/automake/1.16
      target: []
      compiler: []
      version: []
      providers: {}
    bison:
      buildable: false
      externals:
      - spec: bison@2.4.1
        prefix: /usr
      target: []
      providers: {}
      version: []
      compiler: []
    cmake:
      buildable: false
      externals:
      - spec: cmake@3.10.2
        prefix: /apps2/cmake/3.10.2
      target: []
      compiler: []
      version: []
      providers: {}
    findutils:
      buildable: false
      externals:
      - spec: findutils@4.4.2
        prefix: /usr
      target: []
      providers: {}
      version: []
      compiler: []
    freetype:
      buildable: false
      externals:
      - spec: freetype@2.9.1
        prefix: /apps2/freetype/2.9.1
      version: []
      providers: {}
      compiler: []
      target: []
    gmake:
      buildable: false
      externals:
      - spec: gmake@3.81
        prefix: /usr
      target: []
      compiler: []
      version: []
      providers: {}
    libxslt:
      buildable: false
      externals:
      - spec: libxslt@1.1.29
        prefix: /apps2/libxslt/1.1.29
      target: []
      providers: {}
      compiler: []
      version: []
    intltool:
      buildable: false
      externals:
      - spec: intltool@0.51.0
        prefix: /apps2/intltool/0.51.0
      target: []
      providers: {}
      compiler: []
      version: []
    perl:
      buildable: false
      externals:
      - spec: perl@5.24.1
        prefix: /apps2/perl/5.24.1
      providers: {}
      version: []
      compiler: []
      target: []
    rust:
      buildable: false
      externals:
      - spec: rust@1.46.0
        prefix: /home/pan14001admin/.rustup/toolchains/stable-x86_64-unknown-linux-gnu
      target: []
      providers: {}
      compiler: []
      version: []
    zlib:
      buildable: false
      externals:
      - spec: zlib@1.2.8
        prefix: /apps2/zlib/1.2.8
      compiler: []
      providers: {}
      target: []
      version: []
  view: false
```

### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/5264674/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/5264675/spack-build-env.txt)

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by hdezmora** [comment]
Any hint to build openbabel 3.1.1 on CentOS7?  I'm getting these errors

==> Warning: Skipping build of pango-1.50.7-uaoorio4iiwzsc3wo4e5f5puan2wr7y4 since harfbuzz-4.2.1-cwxo6yo2g2kbqoj7jlihyz2sr47rsp67 failed
==> Removing failure mark on rapidjson-1.2.0-2022-03-09-xlfiil2yz2ccaz7tik5qb4hiputy6xn3
==> Removing failure mark on openbabel-3.1.1-uubs2xuqo5mzewjbfxub2dc2fharvgjr
==> Removing failure mark on harfbuzz-4.2.1-cwxo6yo2g2kbqoj7jlihyz2sr47rsp67
==> Removing failure mark on pango-1.50.7-uaoorio4iiwzsc3wo4e5f5puan2wr7y4
==> Error: openbabel-3.1.1-uubs2xuqo5mzewjbfxub2dc2fharvgjr: Package was not installed
==> Error: Installation request failed.  Refer to reported errors for failing package(s).

after mirroring locally openbabel as it doesn't show up on the Spack mirrors but only version 3.0.0.

Any help greatly appreciated!

**Comment by hdezmora** [comment]
Last entries on the build log:

tailf /tmp/spack/spack-stage/spack-stage-harfbuzz-4.2.1-cwxo6yo2g2kbqoj7jlihyz2sr47rsp67/spack-build-out.txt
../src/hb-ot-layout-gpos-table.hh:2846:19:   required from 'typename context_t::return_t OT::PosLookupSubTable::dispatch(context_t*, unsigned int, Ts&& ...) const [with context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
../src/hb-ot-layout-common.hh:1278:118:   required from 'typename context_t::return_t OT::Lookup::dispatch(context_t*, Ts&& ...) const [with TSubTable = OT::PosLookupSubTable; context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
../src/hb-ot-layout-gpos-table.hh:2942:67:   required from 'typename context_t::return_t OT::PosLookup::dispatch(context_t*, Ts&& ...) const [with context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
../src/hb-ot-layout-gsubgpos.hh:3627:5:   required from 'void OT::hb_ot_layout_lookup_accelerator_t::init(const TLookup&) [with TLookup = OT::PosLookup]'
../src/hb-ot-layout-gsubgpos.hh:3948:2:   required from 'OT::GSUBGPOS::accelerator_t<T>::accelerator_t(hb_face_t*) [with T = OT::GPOS; hb_face_t = hb_face_t]'
../src/hb-ot-layout-gpos-table.hh:3110:67:   required from here
../src/hb-vector.hh:312:48: error: 'realloc_vector' was not declared in this scope
       new_array = realloc_vector (new_allocated);
                                                ^
ninja: build stopped: subcommand failed.


Building using this command:

spack install -v openbabel@3.1.1%gcc@4.8.5~ipo+python build_type=RelWithDebInfo arch=linux-centos7-x86_64

Tried also with GCC 9.1.0 same results.

**Comment by omsai** [comment]
> Last entries on the build log:
> 
> ```console
> $ tailf /tmp/spack/spack-stage/spack-stage-harfbuzz-4.2.1-cwxo6yo2g2kbqoj7jlihyz2sr47rsp67/spack-build-out.txt
> ../src/hb-ot-layout-gpos-table.hh:2846:19: required from 'typename context_t::return_t OT::PosLookupSubTable::dispatch(context_t*, unsigned int, Ts&& ...) const [with context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
> ../src/hb-ot-layout-common.hh:1278:118: required from 'typename context_t::return_t OT::Lookup::dispatch(context_t*, Ts&& ...) const [with TSubTable = OT::PosLookupSubTable; context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
> ../src/hb-ot-layout-gpos-table.hh:2942:67: required from 'typename context_t::return_t OT::PosLookup::dispatch(context_t*, Ts&& ...) const [with context_t = OT::hb_get_subtables_context_t; Ts = {}; typename context_t::return_t = hb_empty_t]'
> ../src/hb-ot-layout-gsubgpos.hh:3627:5: required from 'void OT::hb_ot_layout_lookup_accelerator_t::init(const TLookup&) [with TLookup = OT::PosLookup]'
> ../src/hb-ot-layout-gsubgpos.hh:3948:2: required from 'OT::GSUBGPOS::accelerator_t::accelerator_t(hb_face_t*) [with T = OT::GPOS; hb_face_t = hb_face_t]'
> ../src/hb-ot-layout-gpos-table.hh:3110:67: required from here
> ../src/hb-vector.hh:312:48: error: 'realloc_vector' was not declared in this scope
> new_array = realloc_vector (new_allocated);
> ^
> ninja: build stopped: subcommand failed.
> ```
> 
> Building using this command:
> 
>     spack install -v openbabel@3.1.1%gcc@4.8.5~ipo+python build_type=RelWithDebInfo arch=linux-centos7-x86_64
> 
> Tried also with GCC 9.1.0 same results.

@hdezmora In your log, this build failure happens with harfbuzz and not openbabel so the this issue isn't the right place for that report.  The version of harfbuzz was bumped from the 4.2.1 version you tried to 5.1.0 on September 20.  If the fault still persists on your system after updating spack to the new version of harfbuzz, I suggest opening a new issue for harfbuzz and filing the full bug report.  Thank you for reporting!

**Comment by omsai** [comment]
Fixed by #31456 for `openbabel@3.0.0 ~gui` 

```console
$ spack debug report
* **Spack:** 0.19.0.dev0 (a520a7ef284ca7788f1c26d0bb26ea901a2080a7)
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-ivybridge
* **Concretizer:** clingo
$ spack spec -I openbabel ~gui
Input spec
--------------------------------
 -   openbabel~gui

Concretized
--------------------------------
 -   openbabel@3.0.0%gcc@7.5.0~gui~ipo+python build_type=RelWithDebInfo arch=linux-ubuntu18.04-ivybridge
 -       ^boost@1.79.0%gcc@7.5.0+atomic+chrono~clanglibcpp~container~context~contract~coroutine+date_time~debug+exception~fiber+filesystem+graph~graph_parallel~icu+iostreams~json+locale+log+math~mpi+multithreaded~nowide~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded~stacktrace+system~taggedlayout+test+thread+timer~type_erasure~versionedlayout+wave cxxstd=98 patches=a440f96 visibility=hidden arch=linux-ubuntu18.04-ivybridge
[+]          ^bzip2@1.0.8%gcc@7.5.0~debug~pic+shared arch=linux-ubuntu18.04-ivybridge
[+]          ^zlib@1.2.12%gcc@7.5.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu18.04-ivybridge
[+]      ^cairo@1.16.0%gcc@7.5.0~X+fc+ft+gobject+pdf~png~svg patches=7c4da77 arch=linux-ubuntu18.04-ivybridge
[+]          ^fontconfig@2.13.94%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^font-util@1.3.2%gcc@7.5.0 fonts=encodings,font-adobe-100dpi,font-adobe-75dpi,font-adobe-utopia-100dpi,font-adobe-utopia-75dpi,font-adobe-utopia-type1,font-alias,font-arabic-misc,font-bh-100dpi,font-bh-75dpi,font-bh-lucidatypewriter-100dpi,font-bh-lucidatypewriter-75dpi,font-bh-type1,font-bitstream-100dpi,font-bitstream-75dpi,font-bitstream-speedo,font-bitstream-type1,font-cronyx-cyrillic,font-cursor-misc,font-daewoo-misc,font-dec-misc,font-ibm-type1,font-isas-misc,font-jis-misc,font-micro-misc,font-misc-cyrillic,font-misc-ethiopic,font-misc-meltho,font-misc-misc,font-mutt-misc,font-schumacher-misc,font-screen-cyrillic,font-sun-misc,font-winitzki-cyrillic,font-xfree86-type1 arch=linux-ubuntu18.04-ivybridge
[+]              ^freetype@2.11.1%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]                  ^libpng@1.6.37%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^libxml2@2.9.12%gcc@7.5.0~python arch=linux-ubuntu18.04-ivybridge
[+]                  ^libiconv@1.16%gcc@7.5.0 libs=shared,static arch=linux-ubuntu18.04-ivybridge
[+]                  ^xz@5.2.5%gcc@7.5.0~pic libs=shared,static arch=linux-ubuntu18.04-ivybridge
[+]              ^util-linux-uuid@2.37.4%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]          ^glib@2.72.1%gcc@7.5.0~libmount patches=b3fd450 tracing=none arch=linux-ubuntu18.04-ivybridge
[+]              ^gettext@0.21%gcc@7.5.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-ivybridge
[+]                  ^ncurses@6.2%gcc@7.5.0~symlinks+termlib abi=none arch=linux-ubuntu18.04-ivybridge
[+]                  ^tar@1.34%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^libffi@3.4.2%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^pcre@8.45%gcc@7.5.0~jit+multibyte+utf arch=linux-ubuntu18.04-ivybridge
[+]              ^perl@5.34.1%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu18.04-ivybridge
[+]                  ^berkeley-db@18.1.40%gcc@7.5.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu18.04-ivybridge
[+]                  ^gdbm@1.19%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]                      ^readline@8.1%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^python@3.9.12%gcc@7.5.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu18.04-ivybridge
[+]                  ^expat@2.4.8%gcc@7.5.0+libbsd arch=linux-ubuntu18.04-ivybridge
[+]                      ^libbsd@0.11.5%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]                          ^libmd@1.0.4%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]                  ^openssl@1.1.1n%gcc@7.5.0~docs~shared certs=system arch=linux-ubuntu18.04-ivybridge
[+]                  ^sqlite@3.37.2%gcc@7.5.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu18.04-ivybridge
[+]          ^pixman@0.40.0%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]      ^cmake@3.23.1%gcc@7.5.0~doc+ncurses+ownlibs~qt build_type=Release arch=linux-ubuntu18.04-ivybridge
 -       ^eigen@3.4.0%gcc@7.5.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu18.04-ivybridge
[+]      ^libsm@1.2.3%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]          ^libice@1.0.9%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^xproto@7.0.31%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]              ^xtrans@1.3.5%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]      ^pango@1.42.0%gcc@7.5.0~X arch=linux-ubuntu18.04-ivybridge
[+]          ^fribidi@1.0.5%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
[+]          ^gobject-introspection@1.56.1%gcc@7.5.0 patches=6f90bb2,7700828 arch=linux-ubuntu18.04-ivybridge
[+]          ^harfbuzz@2.9.1%gcc@7.5.0~graphite2 arch=linux-ubuntu18.04-ivybridge
[+]              ^icu4c@67.1%gcc@7.5.0 cxxstd=11 arch=linux-ubuntu18.04-ivybridge
[+]      ^pkgconf@1.8.0%gcc@7.5.0 arch=linux-ubuntu18.04-ivybridge
 -       ^rapidjson@1.2.0-2022-03-09%gcc@7.5.0~doc~ipo build_type=RelWithDebInfo patches=a4316c5 arch=linux-ubuntu18.04-ivybridge
 - 
$ spack install openbabel ~gui
...
==> openbabel: Successfully installed openbabel-3.0.0-qhrisfjvoarnld4l6baggxor6mrxegnf
  Fetch: 9.38s.  Build: 2m 58.04s.  Total: 3m 7.41s.
[+] /home/pnanda/spack/opt/spack/linux-ubuntu18.04-ivybridge/gcc-7.5.0/openbabel-3.0.0-qhrisfjvoarnld4l6baggxor6mrxegnf
$ spack load openbabel
$ obabel
No input file or format spec or possibly a misplaced option.
Most options must come after the input files. (-i -o -O -m can be anywhwere.)

Open Babel 3.0.0 -- Oct  6 2022 -- 14:50:46
Usage:
obabel [-i<input-type>] <infilename> [-o<output-type>] -O<outfilename> [Options]
Try  -H option for more information.
```

I that couldn't use the original machine on which I found the fault because spack bootstrap keeps trying to use the operating system `gcc@4.4.7` and failing because clingo needs a C++14 compiler even though the only recognized compiler from `spack compiler list` is `gcc@9.2.0`.
